### ISSUE TITLE
Unable to use mongodb+srv URI format

### ISSUE THREAD

**Issue by Jean85** [mongodb/mongo-php-driver#968]
### Description
I'm unable to use the URI string format `mongo+srv` to connect to my MongoDB cluster.

### Environment
Driver (1.4.2) and ext-mongodb (1.5.3) are latest versions available.
PHP is a Docker container, starting from the official Alpine image:
```
$ php -v
PHP 7.3.3 (cli) (built: Mar  9 2019 01:29:50) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.3, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.3, Copyright (c) 1999-2018, by Zend Technologies
    with Xdebug v2.7.0, Copyright (c) 2002-2019, by Derick Rethans
    with blackfire v1.24.4~linux-musl-x64-non_zts73, https://blackfire.io, by Blackfire
```
```
$ php -i | grep -E 'mongodb|libmongoc|libbson' 

/usr/local/etc/php/conf.d/docker-php-ext-mongodb.ini,
mongodb
libbson bundled version => 1.13.0
libmongoc bundled version => 1.13.0
libmongoc SSL => enabled
libmongoc SSL library => OpenSSL
libmongoc crypto => enabled
libmongoc crypto library => libcrypto
libmongoc crypto system profile => disabled
libmongoc SASL => enabled
libmongoc ICU => disabled
libmongoc compression => enabled
libmongoc compression snappy => disabled
libmongoc compression zlib => enabled
mongodb.debug => no value => no value
```

### Test Script
```
<?php

require_once __DIR__ . '/vendor/autoload.php';

new \MongoDB\Client('mongodb+srv://mongo/');
```

### Expected and Actual Behavior
Expected: no errors.
Actual:
```
PHP Fatal error:  Uncaught MongoDB\Driver\Exception\InvalidArgumentException: Failed to parse MongoDB URI: 'mongodb+srv://mongo/'. Invalid service name in URI. in /var/www/insight/core/vendor/mongodb/mongodb/src/Client.php:87
Stack trace:
#0 /var/www/insight/core/vendor/mongodb/mongodb/src/Client.php(87): MongoDB\Driver\Manager->__construct('mongodb+srv://m...', Array, Array)
#1 /var/www/insight/core/test.php(5): MongoDB\Client->__construct('mongodb+srv://m...')
#2 {main}
  thrown in /var/www/insight/core/vendor/mongodb/mongodb/src/Client.php on line 87
```

**Comment by Jean85** [comment]
Is it correct to say that this is an unsupported feature?
I've found this ticked where it's marked as 1.6.0: https://jira.mongodb.org/browse/PHPC-1292

**Comment by jmikola** [comment]
The error you're seeing originates in [`mongoc_uri_parse_before_slash()`](https://github.com/mongodb/mongo-c-driver/blob/1.13.0/src/libmongoc/src/mongoc/mongoc-uri.c#L1057) within libmongoc after [`mongoc_uri_parse_srv()`](https://github.com/mongodb/mongo-c-driver/blob/1.13.0/src/libmongoc/src/mongoc/mongoc-uri.c#L415) fails. Just by looking at your connection string, I can tell you that "mongo" is an invalid hostname for use with the `mongodb+srv` scheme.

The [SRV specification](https://github.com/mongodb/specifications/blob/master/source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.rst#seedlist-discovery) explains this in a bit more detail:

> In this preprocessing step, the driver will query the DNS server for SRV records on `{hostname}.{domainname}`, prefixed with `_mongodb._tcp.`: `_mongodb._tcp.{hostname}.{domainname}`.
>
> A driver MUST verify that in addition to the `{hostname}`, the `{domainname}` consists of at least two parts: the domain name, and a TLD. Drivers MUST raise an error and MUST NOT contact the DNS server to obtain SRV (or TXT records) if the full URI does not consists of at least three parts.

The SRV examples in the [Connection String](https://docs.mongodb.com/manual/reference/connection-string/#dns-seedlist-connection-format) section of the MongoDB manual satisfy this requirement; however, the text doesn't actually state the two (or more) dot requirement. I went searching for a clear explanation of the requirement in the [SRV record Wikipedia article]() and [RFC2782]() but didn't turn up anything. I have a feeling that the requirement can be inferred from some definitions in the RFC text, but that's still not helpful for the average user. Moreover, users shouldn't be expected to read driver specifications.

I opened [DOCS-12574](https://jira.mongodb.org/browse/DOCS-12574) to request an addition to the MongoDB manual to better explain this requirement.

> Is it correct to say that this is an unsupported feature?

Support for `mongodb+srv` was originally implemented in [PHPC-1015](https://jira.mongodb.org/browse/PHPC-1015), which was released in PHP driver 1.4.0.

[PHPC-1292](https://jira.mongodb.org/browse/PHPC-1292) pertains to polling SRV records for sharded deployments over time. Currently, drivers resolve SRV records once when a client is constructed. That ticket is not yet available, but is scheduled to be included in the upcoming 1.6.0 release.

**Comment by jmikola** [comment]
> I have a feeling that the requirement can be inferred from some definitions in the RFC text, but that's still not helpful for the average user.

I believe I misspoke here. The two (or more) dot requirement does not appear to be a requirement of [RFC2782](https://github.com/mongodb/mongo-php-driver/issues/968); however, it _is_ a requirement for MongoDB's usage of SRV records (likely for security reasons, as noted in my last comment in [DOCS-12574](https://jira.mongodb.org/browse/DOCS-12574)). In any event, this is hopefully something we can clear up in the manual text.

**Comment by Jean85** [comment]
Thanks for the input @jmikola. 

I've refined my test, and I've found that the issue is still there in a corner case, which unfortunately is OUR use case: we are using Consul, so our hostname doesn't end with a public TLD, but with `.consul`. Furthermore, to avoid a DNS resolving issue which seems derived from Alpine, we need to also add the dot at the very end of the address, the top TLD, and this fails.

To summarize:
 - `mongodb+srv://mongodb.example.consul/` works
 - `mongodb+srv://mongodb.example.consul./` does not work.

Do you think that could be patched?

**Comment by jmikola** [comment]
> Do you think that could be patched?

The relevant code to patch would be [`valid_hostname()`](https://github.com/mongodb/mongo-c-driver/blob/1.13.0/src/libmongoc/src/mongoc/mongoc-uri.c#L116), which is called by `mongoc_uri_parse_srv()` (and other functions).

I don't think we can reasonably modify this behavior in the driver proper, as it would conflict with this bit of the [SRV specification](https://github.com/mongodb/specifications/blob/master/source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.rst#seedlist-discovery):

> A driver MUST verify that in addition to the {hostname}, the {domainname} consists of at least two parts: the domain name, and a TLD. Drivers MUST raise an error and MUST NOT contact the DNS server to obtain SRV (or TXT records) if the full URI does not consists of at least three parts.
>
> A driver MUST verify that the host names returned through SRV records have the same parent {domainname}. Drivers MUST raise an error and MUST NOT initiate a connection to any returned host name which does not share the same {domainname}

Even if `valid_hostname()` or `mongoc_uri_parse_srv()` was changed to allow a trailing dot, I would think that the actual host string in the SRV/TXT record would _not_ have a dot (correct me if I'm wrong) and therefore be invalidated due to having a different domain name from the original resolved host string. AFAIK, that behavior is also in place for security reasons to protect against blatant DNS spoofing (e.g. `foo.example.com` resolution yields a SRV record to `mitm-attack.com`).

I'm have very little experience with Alpine (we've had just a handful of issues using that distro over the years) and have never worked with Consul. Is the issue you're describing related to [this write-up](https://github.com/gliderlabs/docker-alpine/blob/master/docs/caveats.md#dns)? If so, I wonder if the suggestion there of running a local caching DNS server might help.

----

To quote [section 3.2.2 of RFC3986](https://tools.ietf.org/html/rfc3986#page-21):

> Such a name consists of a sequence of domain labels separated by ".", each domain label starting and ending with an alphanumeric character and possibly also containing "-" characters.  The rightmost domain label of a fully qualified domain name in DNS may be followed by a single "." and should be if it is necessary to distinguish between the complete domain name and some local domain.

Between this and the grammar definitions in [Collected ABNF for URI](https://tools.ietf.org/html/rfc3986#appendix-A), it looks like a trailing . _is_ permitted in a valid host string. If this might actually address your problem (per "correct me if I'm wrong" above), then I'd be happy to raise the issue internally with other drivers and consider a change to libmongoc.

Offhand, I'm not sure if libmongoc is alone in disallowing trailing dots in its `valid_hostname()` function as this is not explicitly addressed in the [SRV](https://github.com/mongodb/specifications/blob/master/source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.rst) or [Connection String](https://github.com/mongodb/specifications/blob/master/source/connection-string/connection-string-spec.rst) specs. I will note that the [Connection String: Host](https://github.com/mongodb/specifications/blob/master/source/connection-string/connection-string-spec.rst#host) section does reference RFC3986 as the authority on host strings, so it's possible that alone could justify changing libmongoc behavior hear.

**Comment by jmikola** [comment]
FYI: I opened [CDRIVER-3043](https://jira.mongodb.org/browse/CDRIVER-3043) to discuss changing `valid_hostname()` in light of RFC3986. That function dates back to SRV implementation in https://github.com/mongodb/mongo-c-driver/commit/a625b39a2ef868c879b9a6b5eacd2697ddc9a772 so it's possible it was an intentional design restriction. I will aim to follow up once we get to the bottom of it, but feel free to also watch that issue for updates.

**Comment by Jean85** [comment]
I started watching the issue, thank you very much!
 * per "correct me if I'm wrong" above, yes, you're right, the DNS result doesn't have the trailing comma
 * the linked solution from the Alpine docs are not feasible for us: we deploy multiple apps with multiple containers in a Kubernetes cluster, we would have to put dnsmasq inside each container, or replace/proxy the whole consul system..

Since the issue seems to be rooted in the domain comparison, trimming the rightmost dot would be a feasible fix, right?

**Comment by jmikola** [comment]
> Since the issue seems to be rooted in the domain comparison, trimming the rightmost dot would be a feasible fix, right?

My proposed fix was to have libmongoc's `valid_hostname()` function accept a trailing dot in a hostname.

I don't think we can blindly trim a trailing dot from hostname strings, as it may be significant (the RFC talks about using it for local domains). Moreover, I think the SRV logic in the driver would still need to enforce that the SRV result shares a domain with the original string for security reasons (see: [`validate_srv_result()`](https://github.com/mongodb/mongo-c-driver/blob/1.13.0/src/libmongoc/src/mongoc/mongoc-uri.c#L125). In that function, `service` corresponds to the original host from the URI and the function checks that the host in the SRV result shares the same root (i.e. portion of the string following the first dot in the host) as explained in the comment in the middle of the function.

For `mongodb+srv://mongodb.example.consul/`, the driver would validate that any SRV result ended with `example.consul`. `mongodb+srv://mongodb.example.consul./` would be a bit more flexible and I believe we would only check that SRV results end with `consul.`. The latter seems problematic since you stated that the "DNS result doesn't have the trailing dot" in your last comment. Do your SRV results end with `.example.consul`?

I just want to confirm that you are, for the most part, satisfying the check about having a common "domain" in your service (i.e. URI host string) and hosts returned in the SRV result and that the trailing dot in the service is the only thing throwing a wrench into the works here.

**Comment by Jean85** [comment]
>  don't think we can blindly trim a trailing dot from hostname strings, as it may be significant (the RFC talks about using it for local domains).

How is that significant? That's exactly my use case: I have a local .consul top domain, and I want to avoid that the DNS tries to check that name as under an other local domain (`mongo.example.consul.domain`). With the trailing dot, I avoid that append.

> Do your SRV results end with .example.consul?

Yes they do. I fear that your proposed check would fail in my case. So yes, I would be stuck on that.

**Comment by jmikola** [comment]
> How is that significant?

I'm inferring that a trailing a trailing dot may be significant from the [grammar definitions in RFC2782](https://tools.ietf.org/html/rfc3986#appendix-A). Quoting just the relevant bits:

```
host          = IP-literal / IPv4address / reg-name

reg-name      = *( unreserved / pct-encoded / sub-delims )

pct-encoded   = "%" HEXDIG HEXDIG

unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"

sub-delims    = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
```

That grammar doesn't prohibit a valid host string ending in a dot.

> I have a local .consul top domain, and I want to avoid that the DNS tries to check that name as under an other local domain (mongo.example.consul.domain). With the trailing dot, I avoid that append.

Reading some more about trailing dots lead me to [this SO thread](https://stackoverflow.com/a/19498025/162228) and [Fully qualified domain name: Resolution](https://en.wikipedia.org/wiki/Fully_qualified_domain_name#Resolution), which mentions its use to signify the "root of the DNS tree" and prevent resolvers from processing strings "without a dot as unqualified and automatically append(ing) the system's default domain name and the final dot."

Apologies if I was being obtuse here. I haven't done much work with managing DNS and wasn't aware of the use of trailing dots for this purpose. I gather this is what [section 3.2.2](https://tools.ietf.org/html/rfc3986#page-21) was referring to in my citation within https://github.com/mongodb/mongo-php-driver/issues/968#issuecomment-476883918, although I didn't catch the significance on the first read through it and got caught up on the grammar definitions.

Considering all this, I think it may make sense to allow an optional trailing dot any place with receive a hostname in the event users are denoting the "root of the DNS tree" and then take that into account when comparing the service and any SRV results in [`validate_srv_result()`](https://github.com/mongodb/mongo-c-driver/blob/1.13.0/src/libmongoc/src/mongoc/mongoc-uri.c#L125). I'll loop in our security team into [CDRIVER-3043](https://jira.mongodb.org/browse/CDRIVER-3043) to confirm that this is a sound approach.

**Comment by jmikola** [comment]
>> Do your SRV results end with `.example.consul`?
> Yes they do. I fear that your proposed check would fail in my case. So yes, I would be stuck on that.

Just to follow up, is it feasible for the SRV records to also include a trialing dot to indicate that they are rooted? If not, is that simply unsupported or would it be likely to cause some other conflict with how Consul operates?

A colleague just linked me to [SERVER-31965](https://jira.mongodb.org/browse/SERVER-31965) regarding a similar issue with the `mongo` shell (since resolved). That will be helpful to consider alongside any planned changes to driver behavior here and may also shed some light on how we might be able to change the logic within `validate_srv_result()`.

**Comment by Jean85** [comment]
Thanks a lot for the thorough research!

As for adding the trailing dot on the SRV result, I'll ask to my colleague who's handling it and report back.

**Comment by Jean85** [comment]
@jmikola I got corrected by my colleague, our DNS in fact HAS the trailing dot, so no blockers in our case for that! :+1:

**Comment by jmikola** [comment]
@Jean85: Fantastic news! In that case, it sounds like [CDRIVER-3043](https://jira.mongodb.org/browse/CDRIVER-3043) is really all you would need addressed on the libmongoc/PHP driver side. Is that correct?

**Comment by Jean85** [comment]
Yes, it will!

**Comment by jmikola** [comment]
Closing, but please follow the linked JIRA ticket above for updates. If that does get fixed in libmongoc 1.16, you'll see it included in the forthcoming PHP driver 1.7.0 release.

**Comment by scrummitch** [comment]
Hope this helps someone that googles this error: Port is not allowed in the hostname section
https://github.com/mongodb/specifications/blob/master/source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.rst#validation-before-querying-dns
